use of undeclared identifier 'TF_TString_ReserveAmortized'; did you mean 'TF_TString_Reserve'?

Hey!

I'm trying to run [arpieb/deep_learning_elixir](https://github.com), but I'm running into the following error:

```
* Getting nx (https://github.com/elixir-nx/nx.git)
remote: Enumerating objects: 10138, done.        
remote: Counting objects: 100% (2534/2534), done.        
remote: Compressing objects: 100% (1154/1154), done.        
remote: Total 10138 (delta 1554), reused 2173 (delta 1291), pack-reused 7604        
origin/HEAD set to main
* Getting exla (https://github.com/elixir-nx/nx.git)
remote: Enumerating objects: 10138, done.        
remote: Counting objects: 100% (2534/2534), done.        
remote: Compressing objects: 100% (1159/1159), done.        
remote: Total 10138 (delta 1555), reused 2171 (delta 1286), pack-reused 7604        
origin/HEAD set to main
* Getting axon (https://github.com/elixir-nx/axon.git - origin/main)
remote: Enumerating objects: 1855, done.        
remote: Counting objects: 100% (1031/1031), done.        
remote: Compressing objects: 100% (541/541), done.        
remote: Total 1855 (delta 726), reused 732 (delta 480), pack-reused 824        
* Getting nltex (https://github.com/arpieb/nltex.git - origin/master)
remote: Enumerating objects: 211, done.        
remote: Counting objects: 100% (211/211), done.        
remote: Compressing objects: 100% (138/138), done.        
remote: Total 211 (delta 85), reused 177 (delta 53), pack-reused 0        
Resolving Hex dependencies...
Dependency resolution completed:
New:
  elixir_make 0.6.2
  kino 0.3.0
  scidata 0.1.2
  table_rex 3.1.1
  vega_lite 0.1.2
  xla 0.2.0
* Getting scidata (Hex package)
* Getting kino (Hex package)
* Getting vega_lite (Hex package)
* Getting table_rex (Hex package)
* Getting xla (Hex package)
* Getting elixir_make (Hex package)
==> scidata
Compiling 6 files (.ex)
Generated scidata app
==> vega_lite
Compiling 5 files (.ex)
Generated vega_lite app
==> kino
Compiling 13 files (.ex)
Generated kino app
==> nx
Compiling 20 files (.ex)
Generated nx app
==> elixir_make
Compiling 1 file (.ex)
Generated elixir_make app
==> xla
Compiling 2 files (.ex)
Generated xla app
==> exla
Unpacking /Users/philipbrown/Library/Caches/xla/0.2.0/cache/download/xla_extension-x86_64-darwin-cpu.tar.gz into /Users/philipbrown/Library/Caches/mix/installs/elixir-1.12.2-erts-12.0.3/c66f505ce76ecf389af4eb24ccb09ea5/deps/exla/exla/cache
mkdir -p /Users/philipbrown/Library/Caches/mix/installs/elixir-1.12.2-erts-12.0.3/c66f505ce76ecf389af4eb24ccb09ea5/_build/dev/lib/exla/priv
ln -sf /Users/philipbrown/Library/Caches/mix/installs/elixir-1.12.2-erts-12.0.3/c66f505ce76ecf389af4eb24ccb09ea5/deps/exla/exla/cache/xla_extension/lib /Users/philipbrown/Library/Caches/mix/installs/elixir-1.12.2-erts-12.0.3/c66f505ce76ecf389af4eb24ccb09ea5/_build/dev/lib/exla/priv/lib
c++ -fPIC -I/usr/local/Cellar/erlang/24.0.5/lib/erlang/erts-12.0.3/include -isystem cache/xla_extension/include -O3 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-comment -shared -std=c++14 c_src/exla/exla.cc c_src/exla/exla_nif_util.cc c_src/exla/exla_client.cc -o /Users/philipbrown/Library/Caches/mix/installs/elixir-1.12.2-erts-12.0.3/c66f505ce76ecf389af4eb24ccb09ea5/_build/dev/lib/exla/priv/libexla.so -L/Users/philipbrown/Library/Caches/mix/installs/elixir-1.12.2-erts-12.0.3/c66f505ce76ecf389af4eb24ccb09ea5/_build/dev/lib/exla/priv/lib -lxla_extension -flat_namespace -undefined suppress
In file included from c_src/exla/exla.cc:3:
In file included from c_src/exla/exla_nif_util.h:13:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/types.h:22:
In file included from cache/xla_extension/include/tensorflow/core/framework/numeric_types.h:27:
In file included from cache/xla_extension/include/tensorflow/core/platform/types.h:23:
cache/xla_extension/include/tensorflow/core/platform/tstring.h:520:3: error: use of undeclared identifier 'TF_TString_ReserveAmortized'; did you mean 'TF_TString_Reserve'?
  TF_TString_ReserveAmortized(&tstr_, new_size);
  ^
/usr/local/include/tensorflow/core/platform/ctstring.h:45:13: note: 'TF_TString_Reserve' declared here
inline void TF_TString_Reserve(TF_TString *str, size_t new_cap);
            ^
In file included from c_src/exla/exla.cc:4:
c_src/exla/exla_client.h:29:20: warning: private field 'device_' is not used [-Wunused-private-field]
  xla::PjRtDevice* device_;
                   ^
c_src/exla/exla_client.h:30:15: warning: private field 'client_' is not used [-Wunused-private-field]
  ExlaClient* client_;
              ^
2 warnings and 1 error generated.
In file included from c_src/exla/exla_nif_util.cc:1:
In file included from c_src/exla/exla_nif_util.h:13:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/types.h:22:
In file included from cache/xla_extension/include/tensorflow/core/framework/numeric_types.h:27:
In file included from cache/xla_extension/include/tensorflow/core/platform/types.h:23:
cache/xla_extension/include/tensorflow/core/platform/tstring.h:520:3: error: use of undeclared identifier 'TF_TString_ReserveAmortized'; did you mean 'TF_TString_Reserve'?
  TF_TString_ReserveAmortized(&tstr_, new_size);
  ^
/usr/local/include/tensorflow/core/platform/ctstring.h:45:13: note: 'TF_TString_Reserve' declared here
inline void TF_TString_Reserve(TF_TString *str, size_t new_cap);
            ^
1 error generated.
In file included from c_src/exla/exla_client.cc:1:
In file included from c_src/exla/exla_client.h:8:
In file included from c_src/exla/exla_nif_util.h:13:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/types.h:22:
In file included from cache/xla_extension/include/tensorflow/core/framework/numeric_types.h:27:
In file included from cache/xla_extension/include/tensorflow/core/platform/types.h:23:
cache/xla_extension/include/tensorflow/core/platform/tstring.h:520:3: error: use of undeclared identifier 'TF_TString_ReserveAmortized'; did you mean 'TF_TString_Reserve'?
  TF_TString_ReserveAmortized(&tstr_, new_size);
  ^
/usr/local/include/tensorflow/core/platform/ctstring.h:45:13: note: 'TF_TString_Reserve' declared here
inline void TF_TString_Reserve(TF_TString *str, size_t new_cap);
            ^
c_src/exla/exla_client.cc:21:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  buffer_->BlockHostUntilReady();
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from c_src/exla/exla_client.cc:1:
c_src/exla/exla_client.h:29:20: warning: private field 'device_' is not used [-Wunused-private-field]
  xla::PjRtDevice* device_;
                   ^
c_src/exla/exla_client.h:30:15: warning: private field 'client_' is not used [-Wunused-private-field]
  ExlaClient* client_;
              ^
3 warnings and 1 error generated.
make: *** [/Users/philipbrown/Library/Caches/mix/installs/elixir-1.12.2-erts-12.0.3/c66f505ce76ecf389af4eb24ccb09ea5/_build/dev/lib/exla/priv/libexla.so] Error 1
could not compile dependency :exla, "mix compile" failed. You can recompile this dependency with "mix deps.compile exla", update it with "mix deps.update exla" or clean it with "mix deps.clean exla"
```

Hey there! I hit the same error while testing `axon_onnx`, i checked out this repo at 6259450 and `mix test` responds with the following
```
[lines elided]

==> xla
Compiling 2 files (.ex)
Generated xla app

17:15:03.180 [info]  Found a matching archive (xla_extension-x86_64-darwin-cpu.tar.gz), going to download it
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   677  100   677    0     0   4701      0 --:--:-- --:--:-- --:--:--  4734
100 78.1M  100 78.1M    0     0  7819k      0  0:00:10  0:00:10 --:--:-- 9805k
17:15:13.581 [info]  Successfully downloaded the XLA archive
==> exla
Unpacking /Users/danielmorandini/Library/Caches/xla/0.2.0/cache/download/xla_extension-x86_64-darwin-cpu.tar.gz into /Users/danielmorandini/src/git.keepinmind.info/video-taxi/auto-gec/nx/exla/cache
mkdir -p /Users/danielmorandini/src/git.keepinmind.info/video-taxi/auto-gec/nx/exla/_build/test/lib/exla/priv
ln -sf /Users/danielmorandini/src/git.keepinmind.info/video-taxi/auto-gec/nx/exla/cache/xla_extension/lib /Users/danielmorandini/src/git.keepinmind.info/video-taxi/auto-gec/nx/exla/_build/test/lib/exla/priv/lib
c++ -fPIC -I/usr/local/Cellar/erlang/24.2/lib/erlang/erts-12.2/include -isystem cache/xla_extension/include -O3 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-comment -shared -std=c++14 c_src/exla/exla.cc c_src/exla/exla_nif_util.cc c_src/exla/exla_client.cc -o /Users/danielmorandini/src/git.keepinmind.info/video-taxi/auto-gec/nx/exla/_build/test/lib/exla/priv/libexla.so -L/Users/danielmorandini/src/git.keepinmind.info/video-taxi/auto-gec/nx/exla/_build/test/lib/exla/priv/lib -lxla_extension -flat_namespace -undefined suppress
In file included from c_src/exla/exla.cc:3:
In file included from c_src/exla/exla_nif_util.h:13:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/types.h:22:
In file included from cache/xla_extension/include/tensorflow/core/framework/numeric_types.h:27:
In file included from cache/xla_extension/include/tensorflow/core/platform/types.h:23:
cache/xla_extension/include/tensorflow/core/platform/tstring.h:520:3: error: use of undeclared identifier 'TF_TString_ReserveAmortized'; did you mean 'TF_TString_Reserve'?
  TF_TString_ReserveAmortized(&tstr_, new_size);
  ^
/usr/local/include/tensorflow/core/platform/ctstring.h:45:13: note: 'TF_TString_Reserve' declared here
inline void TF_TString_Reserve(TF_TString *str, size_t new_cap);
            ^
1 error generated.
In file included from c_src/exla/exla_nif_util.cc:1:
In file included from c_src/exla/exla_nif_util.h:13:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/types.h:22:
In file included from cache/xla_extension/include/tensorflow/core/framework/numeric_types.h:27:
In file included from cache/xla_extension/include/tensorflow/core/platform/types.h:23:
cache/xla_extension/include/tensorflow/core/platform/tstring.h:520:3: error: use of undeclared identifier 'TF_TString_ReserveAmortized'; did you mean 'TF_TString_Reserve'?
  TF_TString_ReserveAmortized(&tstr_, new_size);
  ^
/usr/local/include/tensorflow/core/platform/ctstring.h:45:13: note: 'TF_TString_Reserve' declared here
inline void TF_TString_Reserve(TF_TString *str, size_t new_cap);
            ^
1 error generated.
In file included from c_src/exla/exla_client.cc:1:
In file included from c_src/exla/exla_client.h:8:
In file included from c_src/exla/exla_nif_util.h:13:
In file included from cache/xla_extension/include/tensorflow/compiler/xla/types.h:22:
In file included from cache/xla_extension/include/tensorflow/core/framework/numeric_types.h:27:
In file included from cache/xla_extension/include/tensorflow/core/platform/types.h:23:
cache/xla_extension/include/tensorflow/core/platform/tstring.h:520:3: error: use of undeclared identifier 'TF_TString_ReserveAmortized'; did you mean 'TF_TString_Reserve'?
  TF_TString_ReserveAmortized(&tstr_, new_size);
  ^
/usr/local/include/tensorflow/core/platform/ctstring.h:45:13: note: 'TF_TString_Reserve' declared here
inline void TF_TString_Reserve(TF_TString *str, size_t new_cap);
            ^
1 error generated.
make: *** [/Users/danielmorandini/src/git.keepinmind.info/video-taxi/auto-gec/nx/exla/_build/test/lib/exla/priv/libexla.so] Error 1
** (Mix) Could not compile with "make" (exit status: 2).
You need to have gcc and make installed. Try running the
commands "gcc --version" and / or "make --version". If these programs
are not installed, you will be prompted to install them.
```

I'm on `macOS BigSur 11.6.2`. References are indeed in the core/platform headers as pointed out @jonatanklosko!
```
ctstring.h:48:inline void TF_TString_ReserveAmortized(TF_TString *str, size_t new_cap);
ctstring_internal.h:345:static inline void TF_TString_ReserveAmortized(TF_TString *str,
ctstring_internal.h:380:  TF_TString_ReserveAmortized(dst, dst_size + src_size);
tstring.h:520:  TF_TString_ReserveAmortized(&tstr_, new_size);
```

--- a/exla/Makefile
+++ b/exla/Makefile
@@ -14,7 +14,7 @@ EXLA_SO = $(PRIV_DIR)/libexla.so
 EXLA_LIB_DIR = $(PRIV_DIR)/lib
 
 # Build Flags
-CFLAGS = -fPIC -I$(ERTS_INCLUDE_DIR) -isystem $(XLA_INCLUDE_PATH) -O3 -Wall -Wextra \
+CFLAGS = -fPIC -I$(ERTS_INCLUDE_DIR) -I$(XLA_INCLUDE_PATH) -O3 -Wall -Wextra \
 	 -Wno-unused-parameter -Wno-missing-field-initializers -Wno-comment \
 	 -shared -std=c++14
 

